<a href="https://colab.research.google.com/github/brendanpshea/data-science/blob/main/Data_Science_06_AggregateFunctions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Aggregate Functions and Data Collection Methods

## Introduction to the Minnesota Beer Rating Data Set

The data set we'll be working with contains information about various beers brewed in Minnesota. It's a fascinating collection of data that helps us understand the preferences and characteristics of different beers. Let's break down the columns in this data set:

1.  BeerName: The name of the beer.
2.  Brewery: The name of the brewery that produces the beer.
3.  Style: The style of the beer, such as "Imperial IPA" or "Russian Imperial Stout." This categorizes the beer based on its taste, color, aroma, and other characteristics.
4.  ABV (Alcohol By Volume): This represents the alcohol content in the beer, measured as a percentage. For example, a beer with an ABV of 9.2% contains 9.2% alcohol by volume.
5.  NumRating: The number of ratings the beer has received on Beer Advocate, a popular beer rating website.
6.  AvgRating: The average rating of the beer on a scale from 1 to 5, where a higher rating indicates a more favorable review.



In [1]:
# Importing the Pandas library
import pandas as pd

# Loading the Minnesota beer rating data from the provided CSV file
!wget 'https://github.com/brendanpshea/data-science/raw/main/data/minnesota_beers.csv' -q
minnesota_beers_df = pd.read_csv("minnesota_beers.csv")

# Displaying the first few rows of the data to understand its structure
minnesota_beers_df.head()


BeerName                  Brewery  \
0                     Nillerzzzzz  Forager Brewing Company   
1                    Abrasive Ale    Surly Brewing Company   
2          Barrel-Aged Silhouette      Lift Bridge Brewery   
3                        Darkness    Surly Brewing Company   
4  Darkness - Bourbon Barrel-Aged    Surly Brewing Company   

                     Style   ABV  NumRating  AvgRating  
0  American Imperial Stout  14.0        143       4.58  
1             Imperial IPA   9.2       4828       4.50  
2   Russian Imperial Stout  11.0        551       4.50  
3   Russian Imperial Stout  12.0       4252       4.48  
4   Russian Imperial Stout  12.0        356       4.46

In [2]:
## Display basic information of table
minnesota_beers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   BeerName   100 non-null    object 
 1   Brewery    100 non-null    object 
 2   Style      100 non-null    object 
 3   ABV        100 non-null    float64
 4   NumRating  100 non-null    int64  
 5   AvgRating  100 non-null    float64
dtypes: float64(2), int64(1), object(3)
memory usage: 4.8+ KB


#### Examples:

-   Nillerzzzzz is an "American Imperial Stout" with an ABV of 14.0%, and it has received 143 ratings with an average rating of 4.58.
-   Abrasive Ale is an "Imperial IPA" with an ABV of 9.2%, and it has received 4828 ratings with an average rating of 4.50.

This data set provides a rich source of information for exploring various aspects of the beer industry in Minnesota, including consumer preferences, beer characteristics, and brewery performance. By using aggregate functions in Pandas, we can analyze this data in many interesting ways, such as finding the highest-rated beers, understanding the popularity of different beer styles, or identifying trends in alcohol content.

In the following sections, we'll learn how to use Pandas to perform these analyses and more, providing you with valuable insights and techniques to work with structured data. So grab your favorite beer (if you're of legal drinking age, of course!) and let's dive into the world of data analysis with Pandas!

## Grouping
In this section, we will explore how to group data by a specific category and then calculate the mean, or average, for each group. The emphasis is on the powerful concept of "grouping," which allows us to segment data into categories and perform calculations on each category separately.
### What is the Average Alcohol Content (ABV) for Each Beer Style?

Our first example will be calculating the average Alcohol By Volume (ABV) for each beer style.

Imagine you are at a beer festival in Minnesota, and you are curious to know how different styles of beer compare in terms of alcohol content. Some styles might be stronger, while others might be lighter. How can we figure this out using data analysis?

The answer lies in grouping the beers by their style and then calculating the average ABV for each group. By doing so, we can gain insights into how different styles compare in terms of alcohol content. Here's how we can achieve this:

1.  Group the Beers by Style: First, we'll use the `groupby` method in Pandas to group the beers by their style. This will create a collection of groups where each group contains all the beers of a particular style.

2.  Calculate the Average ABV: Next, we'll use the `mean` function to calculate the average ABV for each group. The mean function adds up all the ABV values in a group and then divides by the number of values, giving us the average.

3.  Analyze the Results: Finally, we'll examine the results to understand how different beer styles compare in terms of alcohol content. This analysis can lead to interesting insights and discussions about beer culture, preferences, and more. We'll use the Pandas method `sort_values()` to give our results meaning.

We'll now use the Minnesota beer rating data set to calculate the average ABV for each beer style. Here's the code to do it, along with the results:

In [3]:
# Grouping the beers by their style and calculating the average ABV for each group
average_abv_by_style = minnesota_beers_df.groupby('Style')['ABV'].mean()

# Sorting the result for better visibility
average_abv_by_style_sorted = average_abv_by_style.sort_values(ascending=False)

# Displaying the result
average_abv_by_style_sorted.head(10)


Style
English Barleywine         14.400000
American Imperial Stout    12.018182
Wheatwine                  11.500000
Russian Imperial Stout     11.070000
American Barleywine         9.900000
Imperial Porter             9.800000
Smoked Porter               9.500000
Belgian Dark Strong Ale     9.500000
Imperial IPA                8.904762
Fruited Kettle Sour         8.000000
Name: ABV, dtype: float64

By grouping the beers by style and calculating the average ABV, we can clearly see how different styles compare. For example, English Barleywine and American Imperial Stout are among the strongest styles in terms of alcohol content, while others, like Imperial IPA and Fruited Kettle Sour, are slightly lighter.  This analysis provides a valuable perspective for beer enthusiasts, brewers, and even regulators who might be interested in understanding the alcohol content of various beer styles.


### How to "Count" Beers by Brewery

If you want to know which breweries produce the most beers, you can group the data by brewery and then count the number of beers for each group. The steps are:

1.  Group Beers by Brewery: Use the `groupby` method to group the beers by the brewery.
2.  Count the Beers: Use the `count` function to count the number of beers for each brewery group.
3.  Analyze the Results: Identify the breweries that produce the most beers. Again, we will use `df.sort_values()`.

We'll now use the Minnesota beer rating data set to count the number of beers produced by each brewery. Here's the code to do it:


In [4]:
# Grouping the beers by brewery and counting the number of beers for each brewery
beers_by_brewery = minnesota_beers_df.groupby('Brewery')['BeerName'].count()

# Sorting the result to find the breweries that produce the most beers
top_breweries_by_beer_count = beers_by_brewery.sort_values(ascending=False)

# Displaying the top 5 breweries
top_breweries_by_beer_count.head(5)

Brewery
Surly Brewing Company         14
Modist Brewing Co.            14
Lupulin Brewing                9
Barrel Theory Beer Company     9
BlackStack Brewing             9
Name: BeerName, dtype: int64

### Understanding Sorting: The `sort_values()` Function

Sorting is the process of arranging data in a specific order, either ascending (from smallest to largest) or descending (from largest to smallest). In Pandas, the `sort_values()` function is used to perform sorting on a Series or DataFrame. In our analysis of the Minnesota Beer data, we used the `sort_values()` function to organize data in a meaningful way. Here's a closer look at how we used this function:

1.  Sorting by a Specific Column: We can sort the DataFrame by a specific column, such as the average rating. This helps us identify the breweries with the highest or lowest ratings.

2.  Ascending and Descending Order: By setting the `ascending` parameter, we can control the order of sorting. For example, `ascending=False` will sort the values in descending order, showing us the breweries with the highest ratings first.

3.  Creating a New Copy: The `sort_values()` function returns a new sorted Series or DataFrame, leaving the original data unchanged. This allows us to maintain the original dataset while working with a sorted version.

**Example:** We sorted the average rating for each brewery in descending order to find the breweries with the highest average ratings:

```python
top_breweries_by_avg_rating = average_rating_by_brewery.sort_values(ascending=False)
```

This code created a new Series with the breweries sorted by their average rating, allowing us to easily identify the top-rated breweries.

Sorting is a fundamental operation in data analysis that helps us organize and interpret data. By understanding how to use the `sort_values()` function in Pandas, we can arrange data in meaningful ways, making it easier to analyze and draw insights. Whether we want to find the highest-rated breweries or understand the distribution of beer styles, sorting provides a powerful tool to enhance our data exploration.

### Exercise

Your task is to find out how many beers are produced for each beer style in the Minnesota beer rating data set. Here's what to do:

1.  Group Beers by Style: Use the `groupby` method to group the beers by their style.
2.  Count the Beers: Use the `count` function to count the number of beers for each style group.
3.  Analyze the Results: Identify the top 3 beer styles based on the number of beers. Write a brief analysis of what you observe.

You can should be able to use the code blocks provided above, with only very small changes (to group by style instead of brewery).

In [7]:
# Grouping the beers by beer style and count the number of beers
# beers_by_style = ?

# Sorting the result to find the breweries that produce the most beers
# top_breweries_by_beer_style = ?

# Displaying the top 3 style. You'll need to use df,head()


### Finding the `max()` Rated Brewery

Imagine you are a beer connoisseur, and you want to visit the brewery with the highest average rating in Minnesota. You have a dataset of the most highly rated beers, and you want to analyze this data to make your decision. Here's how you can approach this task:

1.  Calculate the Average Rating for Each Brewery: First, you need to group the beers by brewery and then calculate the average rating for each group. This will give you a clear picture of how each brewery is rated by beer enthusiasts.

2.  Identify the Brewery with the Highest Average Rating: Next, you need to find the maximum average rating among all the breweries. This will help you identify the brewery that is most favored among the reviewers.

3.  Consider the Context: Since the dataset includes only the most highly rated beers, the analysis will reflect the preferences of those who rate beers highly. It's essential to keep this context in mind when interpreting the results.


We'll now use the Minnesota beer rating data set to find the brewery with the highest average rating. Here's the code to do it:

In [6]:
# Finding the average rating for each brewery
average_rating_by_brewery = minnesota_beers_df.groupby('Brewery')['AvgRating'].mean()

# Identifying the brewery with the highest average rating
# idmax() gets us the actual name of the brewery
highest_avg_rating_brewery = average_rating_by_brewery.idxmax()
# max() gets use the number
highest_avg_rating_value = average_rating_by_brewery.max()

highest_avg_rating_brewery, highest_avg_rating_value


('Lift Bridge Brewery', 4.5)

The brewery with the highest average rating among the most highly rated beers in Minnesota is Lift Bridge Brewery, with an average rating of 4.5. This result tells us that Lift Bridge Brewery is highly favored among reviewers, especially considering that the dataset includes only the most highly rated beers. It's a significant achievement for a brewery to stand out in such a competitive field.

In this example, you'll notice we use two related functions `df.max()` and `df.idmax()`

-   `max`: Returns the numerical value of the maximum element. In our example, it gives us the highest average rating as a number.
-   `idxmax`: Returns the index (or label) of the maximum element. In our example, it gives us the name of the brewery that has the highest average rating.

The combination of the `mean`, `idmax` and `max` functions has allowed us to perform a nuanced analysis of the brewery ratings. By understanding how to calculate the average rating for each brewery and then identify the one with the highest average, we have unlocked insights that could be valuable for consumers, business analysts, and beer enthusiasts.

Remember, the concepts of mean and max are not limited to beer ratings. They are fundamental statistical tools that can be applied across various fields and contexts. Whether you're analyzing customer satisfaction, academic performance, or market trends, understanding how to use these functions can help you make informed decisions and uncover meaningful insights.

Now that you've learned how to use the mean and max functions to analyze brewery ratings, think about how you can apply these concepts to other data sets and questions. The world of data analysis is vast and exciting, and these tools are just the beginning of what you can explore and discover!

### Exercise: Finding the `min()`

Take the code block above (which we used to find the highest rated brewery) and use it to find the **lowest** rated brewery. It's basically the same idea, but you'll be using `min()` and `idmin()` instead of `max()` and `idmax()`.

In [8]:
# Change the code below to use min() and idmin() instead of max and idmax
# Finding the average rating for each brewery
average_rating_by_brewery = minnesota_beers_df.groupby('Brewery')['AvgRating'].mean()
highest_avg_rating_brewery = average_rating_by_brewery.idxmax()
highest_avg_rating_value = average_rating_by_brewery.max()
highest_avg_rating_brewery, highest_avg_rating_value

('Lift Bridge Brewery', 4.5)

## Putting it Altogether: A Comprehensive Report

Now, let's put together what we've learned and create a comprhensive report on each beer style and its associated alchohol content. We'll also meet the `df.agg()` function, which allows us to apply multiple aggregate functions to our data.

In [16]:
# Grouping by beer style and calculating the required statistics for the 'ABV' column
report_abv_by_style = minnesota_beers_df.groupby('Style')['ABV'].agg(
    min_abv='min',
    max_abv='max',
    mean_abv='mean',
    median_abv='median',
    sd_abv='std',
    count_style = 'count'
)

# Displaying the report (showing top 5 rows for brevity)
report_abv_by_style


min_abv  max_abv   mean_abv  median_abv    sd_abv  \
Style                                                                         
American Barleywine           9.9     9.90   9.900000        9.90       NaN   
American Brown Ale            5.1     5.50   5.300000        5.30  0.282843   
American IPA                  6.0     7.90   6.900000        7.00  0.562139   
American Imperial Stout      10.0    14.00  12.018182       12.50  1.483791   
American Lager                5.2     6.50   5.850000        5.85  0.919239   
American Pale Ale             7.5     7.50   7.500000        7.50       NaN   
American Porter               5.3     5.30   5.300000        5.30       NaN   
American Stout                6.0     6.00   6.000000        6.00       NaN   
Belgian Dark Strong Ale       9.5     9.50   9.500000        9.50       NaN   
Berliner Weisse               7.2     7.20   7.200000        7.20       NaN   
Black IPA                     6.0     6.00   6.000000        6.00       NaN   
Czech Pale Lager              4.1     4.10   4.100000        4.10       NaN   
English Barleywine           14.4    14.40  14.400000       14.40       NaN   
Fruited Kettle Sour           8.0     8.00   8.000000        8.00       NaN   
Imperial IPA                  7.7    12.50   8.904762        8.50  1.193095   
Imperial Porter               8.3    11.30   9.800000       10.00  1.148913   
Kellerbier / Zwickelbier      5.1     5.10   5.100000        5.10       NaN   
Milkshake IPA                 7.3     7.30   7.300000        7.30       NaN   
New England IPA               6.2    10.10   7.684211        7.50  1.159149   
Russian Imperial Stout        9.3    12.00  11.070000       11.70  1.178558   
Smoked Beer                   7.0     7.00   7.000000        7.00       NaN   
Smoked Porter                 9.5     9.50   9.500000        9.50       NaN   
Sweet / Milk Stout            6.3     6.30   6.300000        6.30       NaN   
Wheatwine                    11.5    11.50  11.500000       11.50       NaN   
Wild Ale                      6.4     6.66   6.573333        6.66  0.150111   

                          count_style  
Style                                  
American Barleywine                 1  
American Brown Ale                  2  
American IPA                       11  
American Imperial Stout            11  
American Lager                      2  
American Pale Ale                   1  
American Porter                     1  
American Stout                      1  
Belgian Dark Strong Ale             1  
Berliner Weisse                     1  
Black IPA                           1  
Czech Pale Lager                    1  
English Barleywine                  1  
Fruited Kettle Sour                 1  
Imperial IPA                       21  
Imperial Porter                     5  
Kellerbier / Zwickelbier            1  
Milkshake IPA                       1  
New England IPA                    19  
Russian Imperial Stout             10  
Smoked Beer                         1  
Smoked Porter                       1  
Sweet / Milk Stout                  1  
Wheatwine                           1  
Wild Ale                            3

Here, we use the `df.agg()` function in Pandasto aggregate a DataFrame using one or more aggregate functions (of the type we've been learning about). In the code above, the `df.agg()` function is used to group the `minnesota_beers_df` DataFrame by the `Style` column and calculate the minimum, maximum, mean, median, and standard deviation of the ABV values for each beer style. The results of the aggregation are stored in the report_abv_by_style DataFrame.

Here is a more concise breakdown of the code:

- `minnesota_beers_df.groupby('Style')['ABV']:` This line groups the minnesota_beers_df DataFrame by the Style column and selects the ABV column.

- `.agg(min_abv='min', max_abv='max', mean_abv='mean', median_abv='median', sd_abv='std')`: This line applies the aggregation functions to the ABV column. The names of the functions are passed as strings, but you can also pass the functions themselves.

- `report_abv_by_style:` This line stores the results of the aggregation in the report_abv_by_style DataFrame.

- `report_abv_by_style.head()`: This line displays the top 5 rows of the report_abv_by_style DataFrame.

In the output, you'll notice that standard deviation is sometimes "not a number" (`NaN`). This is because standard deviation isn't defined when there is only 1 data item (e.g., only one beer of a given style).

## Introduction to Data Collection Methods

In the dynamic world of data analysis, the process of gathering information is as vital as the insights derived from it. Data serves as the foundation upon which we build our understanding, make decisions, and uncover trends. Whether we are exploring the preferences of beer enthusiasts, analyzing the market trends of breweries, or studying the impact of different brewing techniques, the way we collect data shapes the quality and depth of our analysis.

Different research questions and objectives require diverse data collection methods. From understanding the global reach of different beer brands to identifying local craft beer trends, various techniques can be employed to gather the information needed. In this section, we will delve into some of the most commonly used data collection methods, each offering unique advantages and applications:

1.  Web Scraping: Extracting data from websites to analyze beer ratings, customer reviews, or price trends.
2.  Public Databases: Accessing comprehensive datasets related to beer production, sales, or regulation from governmental or organizational repositories.
3.  Application Programming Interface (API)/Web Services: Connecting to specialized platforms or services that provide data on beer varieties, brewery locations, or industry statistics.
4.  Survey: Conducting targeted questionnaires to gather opinions and preferences from beer consumers, brewers, or retailers.
5.  Sampling: Selecting a representative subset of beer products or consumers to conduct in-depth analysis.
6.  Observation: Observing behavior and patterns in beer consumption, purchasing habits, or brewery operations.

By understanding and mastering these data collection methods, we can paint a vivid picture of the multifaceted world of beer. Whether a novice homebrewer seeking to understand preferences or a multinational corporation analyzing global trends, these techniques empower us to gather the data necessary to make informed decisions and create impactful strategies.

In the following sections, we will explore each of these data collection methods in detail, understanding their applications, advantages, challenges, and how they can be effectively utilized in the context of the beer industry.

## Webscraping: Extracting Beer Data from HTML Pages
**Web scraping** is a vital data collection method that involves extracting information from websites. In the context of our beer analysis, the data frame we used earlier was obtained through web scraping from Beer Advocate's Top Rated Beers in Minnesota (https://www.beeradvocate.com/beer/top-rated/us/mn/). This webpage provides a treasure trove of details on highly rated beers, including the beer's name, style, ABV (Alcohol By Volume), and average rating.

The process of web scraping begins with sending an HTTP request to the target URL, in this case, the Beer Advocate page. Upon receiving the HTML content of the webpage from the server, data scientists use libraries like BeautifulSoup or Scrapy to parse the HTML content and navigate through the elements. By identifying specific HTML tags and attributes, they can extract the desired data, such as beer names, styles, and ratings. Once extracted, this data must be cleaned and structured into a usable format, such as a Pandas DataFrame, by removing unnecessary characters and whitespace. (Note: For the Minnesota Beer dataframe, this is actually pretty involved!).

#### Structure of HTML

Webscraping gets its data from **HTML**, or **HyperText Markup Language**, which is the standard language used to create and design webpages. It defines the structure of a webpage, organizing content into elements such as headings, paragraphs, images, links, and tables. These elements are represented using tags, which encapsulate the content and provide instructions on how it should be displayed in a web browser.


An HTML document is structured as a hierarchy of elements, often visualized as a tree. Here's a basic breakdown of the structure:

1.  Document Type Declaration: `<!DOCTYPE html>` specifies the version of HTML being used.
2.  HTML Element: The root element, `<html>`, contains all other elements on the page.
3.  Head Element: The `<head>` section includes meta information, such as the title and links to stylesheets.
4.  Body Element: The `<body>` section contains the main content of the webpage, organized into various elements.

#### Key Concepts in HTML

-   Tags: HTML uses tags to define elements. Tags usually come in pairs, with an opening tag (`<tag>`) and a closing tag (`</tag>`).
-   Attributes: Elements may have attributes that provide additional information, such as class, id, or style.
-   Hierarchy: Elements are nested within one another, creating a parent-child relationship.

And example of HTML woudl be as follows:

In [20]:
%%html
<!DOCTYPE html>
<html>
<head>
    <title>Top Rated Beers</title>
</head>
<body>
    <h1>Minnesota's Best Beers</h1>
    <p>Explore the <a href="link">top-rated beers</a> in Minnesota.</p>
    <table>
        <!-- Table rows and columns containing beer data -->
        <!--This table is where we would *scrape* data from-->
         <tr> <th>Name</th> <th>Style</th> <th>ABV</th> <th>Rating</th> </tr>
         <tr> <td>North Star Lager</td> <td>Lager</td> <td>4.8%</td> <td>4.2</td> </tr>
         <tr> <td>Lakeview IPA</td> <td>India Pale Ale</td> <td>6.5%</td> <td>4.5</td> </tr>
        <tr> <td>Timberwolf Stout</td> <td>Stout</td> <td>5.2%</td> <td>4.1</td> </tr>
    </table>
</body>
</html>


When web scraping, HTML serves as the primary source of data. By navigating the HTML structure and extracting content from specific tags and attributes, data scientists can retrieve valuable information from webpages. The data extracted can range from text and numbers to links and images, depending on the content of the webpage.

### Webscraping: Strengths and Weaknesses
Web scraping offers immense value to data scientists, particularly when data is not readily available through public databases or APIs. It allows for the customization of data collection, enabling data scientists to extract only the relevant information. This approach is widely used for purposes ranging from competitive analysis, where businesses scrape data from competitors' websites, to trend analysis, where data scientists identify changes in popularity or pricing over time.

However, web scraping is not without challenges and potential sources of error. The HTML structure of a website may change, causing the scraping code to break and require updates. Legal and ethical considerations must always be taken into account, as scraping might be restricted or prohibited by a website's terms of service. Additionally, the data on webpages may contain human errors or inconsistencies, leading to inaccuracies in the scraped information. Websites might also implement rate limiting, which restricts repeated requests from the same IP address, necessitating responsible scraping practices.

In conclusion, web scraping played an indispensable role in our exploration of Minnesota's top-rated beers. It's a technique that, while requiring careful handling and consideration of potential errors, offers a versatile and robust tool for data collection. The insights derived from the data scraped from Beer Advocate's website illustrate how web scraping can fuel research, innovation, and informed decision-making in the world of beer and beyond.

### Questions: Web Scraping
1. Consider the ethical aspects of web scraping. How can you ensure that you're scraping data responsibly and within legal boundaries? What might be the consequences of not adhering to ethical guidelines? For example, our scraping above--a small chunk of data for educational purposes, is allowed. However, how would it change if we (repeatedly) scrapped large amounts of data for commerical purposes?

2. Web scraping can present various challenges, especially with websites that change their structure or block scraping. What common obstacles might you encounter when scraping data, and how would you overcome them to ensure accurate data collection?"

3. Web scraping is used across different industries for various purposes. How might you apply web scraping in your own projects or interests, such as analyzing beer ratings? Can you think of innovative ways to utilize web scraping in other fields or topics that interest you?

### My Answers: Web Scraping
1.

2.

3.

## Public Databases
Public databases are repositories of information that are freely accessible to the general public. They encompass a wide array of data collected by government agencies, research institutions, organizations, and more. An example of such a public database is the one provided by the National Institute on Alcohol Abuse and Alcoholism (NIAAA), which publishes surveillance reports related to alcohol abuse and alcoholism (https://www.niaaa.nih.gov/publications/surveillance-reports). These reports, available on the NIAAA's website, offer valuable insights into trends, demographics, and the health impact of alcohol consumption.

Finding public databases can be achieved through various means. Using search engines with keywords related to the topic, exploring government websites, academic institutions, and data portals like data.gov are common ways to locate these resources. These databases often present data in standard formats such as CSV (Comma-Separated Values) and XLSX (Microsoft Excel format), making them accessible for analysis using tools like Microsoft Excel or Python's Pandas library. Once downloaded, the data can be imported, explored, and analyzed to uncover patterns, correlations, and specific insights.

#### How to Use Public Databases

Public databases typically offer data in common formats such as CSV (Comma-Separated Values) and XLSX (Microsoft Excel format). Here's how to use them:

1.   Locate the desired dataset and download it in the available format.
2.   Use software like Microsoft Excel, Python's Pandas library (as we've done here), or R to read and analyze the data.
3.  Utilize analytical methods to explore trends, patterns, correlations, or any specific questions related to the data.

#### Strengths and Weaknesses of Public Databases

Some strengths of public datbases include:

  -  Usually freely available to anyone.
  -   Often curated and maintained by reputable organizations.
  -   Cover a wide range of subjects, providing valuable insights.

Some potential weaknesses are:

  -   May vary in quality, accuracy, and completeness.
  -   Some databases may offer data in less common or older formats.
  -   Information may become outdated if not regularly updated.

The utilization of public databases comes with notable strengths and weaknesses. Among the strengths are their accessibility, reliability, and coverage of diverse topics. Being usually free and maintained by reputable entities, they provide a rich source of information on a wide variety of subjects. However, they may also present challenges such as varying data quality, format limitations, and potential lack of updates, which can impact the accuracy and relevance of the information.

In [ ]:
# let's download public database to explore!
# This will show up in your downloads, where you can click on it.
!wget "https://www.niaaa.nih.gov/sites/default/files/pcyr1970-2021.txt"


## What is an API?

Write something here.

In [17]:
import requests
import json

# The API endpoint URL
url = "https://api.openbrewerydb.org/v1/breweries?by_state=minnesota&by_city=rochester"

# Make a GET request to the API
response = requests.get(url)

# Check the response status code
if response.status_code == 200:
  # The request was successful, get the data
  data = response.json()

  # Print the first row of data
  print(json.dumps(data[0], indent=2))

else:
  # The request failed, print the error message
  print(response.status_code)
  print(response.text)

{
  "id": "190e7e9c-9e20-4b26-b820-f951c75c27b3",
  "name": "Forager Brewing Company",
  "brewery_type": "brewpub",
  "address_1": "1005 6th St NW",
  "address_2": null,
  "address_3": null,
  "city": "Rochester",
  "state_province": "Minnesota",
  "postal_code": "55901-2741",
  "country": "United States",
  "longitude": "-92.47868101",
  "latitude": "44.02937696",
  "phone": "5072587490",
  "website_url": "http://www.foragerbrewery.com",
  "state": "Minnesota",
  "street": "1005 6th St NW"
}


In [18]:
# Convert to pandas
brewery_df = pd.DataFrame(data)
brewery_df.head()


id                          name  \
0  190e7e9c-9e20-4b26-b820-f951c75c27b3       Forager Brewing Company   
1  0a5882a8-ba47-43a1-b5f5-804fa1ea8f2a  Grand Rounds Brewing Company   
2  90d71e53-16cf-4e32-ab36-fd068ab717aa          Kinney Creek Brewery   
3  2688cb2b-a937-42d7-a49f-3e1894a66021        Little Thistle Brewing   
4  4d1d38b6-5514-45bc-a020-74efd0c1bf8f           LTS Brewing Company   

  brewery_type         address_1 address_2 address_3       city  \
0      brewpub    1005 6th St NW      None      None  Rochester   
1      brewpub       4 3rd St SW      None      None  Rochester   
2        micro    1016 7th St NW      None      None  Rochester   
3     planning              None      None      None  Rochester   
4        micro  2001 32nd Ave NW      None      None  Rochester   

  state_province postal_code        country     longitude     latitude  \
0      Minnesota  55901-2741  United States  -92.47868101  44.02937696   
1      Minnesota  55902-3019  United States  -92.46339522   44.0202771   
2      Minnesota  55901-2668  United States  -92.47808893   44.0312604   
3      Minnesota       55901  United States          None         None   
4      Minnesota  55901-8321  United States   -92.5104808   44.0447412   

        phone                        website_url      state            street  
0  5072587490      http://www.foragerbrewery.com  Minnesota    1005 6th St NW  
1  5072921628  http://www.grandroundsbrewing.com  Minnesota       4 3rd St SW  
2        None                               None  Minnesota    1016 7th St NW  
3  5073584647   http://www.littlethistlebeer.com  Minnesota              None  
4  5072268280          http://www.ltsbrewing.com  Minnesota  2001 32nd Ave NW